### DISCUSSION TITLE
Ways to reduce the calculation time

### DISCUSSION THREAD

**Discussion by LiuPengPeter** [idaholab/moose#16483] [Q&A Modules: General] [answered]
Hello,

In my model, the time_scale is 1.0E+9 (from s to ns). The length dt=1000. So **1.0E+6 of steps** have to be used to get the result of  200s. So huge!

What's more, the situation won't be better if I change the time_scale and dt. (Larger dt will be adapted to around 1000 finally.)

Is there any other methods recommended  to reduce the calculation time? Will scaling factor help?

Thank you a lot!

My execution is in the following:
```
[Executioner]
  type = Transient
  scheme = bdf2
  solve_type = 'PJFNK'
  #########PETSC_OPTIONS###################very_important###################
  petsc_options_iname = '-pc_type -sub_pc_type   -sub_pc_factor_shift_type'
  petsc_options_value = 'asm       ilu            nonzero'
  l_max_its = 50
  l_tol = 1e-4
  nl_max_its = 20
  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-10
  end_time = 2.00E+11
[./TimeStepper]
    type = IterationAdaptiveDT
    dt = 1.00E+03
    cutback_factor = 0.8
    growth_factor = 2.0 
    optimal_iterations = 7
 [../]
[]
[Preconditioning]
  active = 'full'
  [./full]
    type = SMP
    full = true
  [../]
  [./mydebug]
    type = FDP
    full = true
  [../]
[]
```

**Comment by aeslaughter** [comment]
@fdkong Can you help here?

**Comment by fdkong** [comment, marked as answer]
It is often problem dependent.

1) Run the code in parallel may be able to make the code faster

2) If the Jacobian matrix is accurate, use `solve_typ=newton` could speedup a bit 

3) If the number of linear iterations is not small, could increase the fill-in levels `-sub_pc_factor_levels 2`

**Reply by LiuPengPeter** [reply]
It runs very quickly, about 5 seconds for one time step. 
I set the end_time =  1.0E+9, (200 s), but after one day only get the result of 1.0E+7. So maybe 100 days are needed...

The unit of time is ns (10E-9 s) and the simulation won't converge well if I change it. Why is it so small? 

Thank you for your help!

Time Step 20412, time = 3.48711e+07, dt = 2063.65
 0 Nonlinear |R| = 1.301949e-08
    |residual|_2 of individual variables:
                       w:    1.30195e-08
                       c:    4.55567e-14
                       eta1: 9.17821e-19
                       eta2: 7.08598e-20
                       eta3: 2.41848e-15
                       c1:   5.31692e-13
                       c2:   5.29201e-13
                       c3:   3.56361e-14
      0 Linear |R| = 1.301949e-08
      1 Linear |R| = 1.301949e-08
      2 Linear |R| = 1.301947e-08
      3 Linear |R| = 1.301946e-08
      4 Linear |R| = 1.301945e-08
      5 Linear |R| = 1.301943e-08
      6 Linear |R| = 1.301939e-08
      7 Linear |R| = 1.301930e-08
      8 Linear |R| = 1.301916e-08
      9 Linear |R| = 1.301889e-08
     10 Linear |R| = 1.301845e-08
     11 Linear |R| = 1.301758e-08
     12 Linear |R| = 1.301579e-08
     13 Linear |R| = 1.301191e-08
     14 Linear |R| = 1.300521e-08
     15 Linear |R| = 1.299228e-08
     16 Linear |R| = 1.297684e-08
     17 Linear |R| = 1.294398e-08
     18 Linear |R| = 1.290817e-08
     19 Linear |R| = 1.283983e-08
     20 Linear |R| = 1.274005e-08
     21 Linear |R| = 1.255210e-08
     22 Linear |R| = 1.225769e-08
     23 Linear |R| = 1.176580e-08
     24 Linear |R| = 1.112838e-08
     25 Linear |R| = 1.033345e-08
     26 Linear |R| = 9.153863e-09
     27 Linear |R| = 8.003364e-09
     28 Linear |R| = 6.622598e-09
     29 Linear |R| = 5.597964e-09
     30 Linear |R| = 8.052611e-09
     31 Linear |R| = 8.052609e-09
     32 Linear |R| = 8.052323e-09
     33 Linear |R| = 8.052207e-09
     34 Linear |R| = 8.052110e-09
     35 Linear |R| = 8.051696e-09
     36 Linear |R| = 8.051194e-09
     37 Linear |R| = 8.049746e-09
     38 Linear |R| = 8.047512e-09
     39 Linear |R| = 8.043057e-09
     40 Linear |R| = 8.036283e-09
     41 Linear |R| = 8.022776e-09
     42 Linear |R| = 7.995286e-09
     43 Linear |R| = 7.937876e-09
     44 Linear |R| = 7.842004e-09
     45 Linear |R| = 7.663963e-09
     46 Linear |R| = 7.471331e-09
     47 Linear |R| = 7.100422e-09
     48 Linear |R| = 6.752856e-09
     49 Linear |R| = 6.194370e-09
     50 Linear |R| = 5.573689e-09
  Linear solve did not converge due to DIVERGED_ITS iterations 50
 1 Nonlinear |R| = 9.034513e-09
    |residual|_2 of individual variables:
                       w:    8.13015e-10
                       c:    1.20437e-09
                       eta1: 3.7189e-15
                       eta2: 6.26328e-18
                       eta3: 3.23551e-12
                       c1:   5.46881e-09
                       c2:   5.46157e-09
                       c3:   4.44682e-09
      0 Linear |R| = 9.034513e-09
      1 Linear |R| = 9.034496e-09
      2 Linear |R| = 9.033128e-09
      3 Linear |R| = 9.032453e-09
      4 Linear |R| = 9.031768e-09
      5 Linear |R| = 9.030367e-09
      6 Linear |R| = 9.027351e-09
      7 Linear |R| = 9.020515e-09
      8 Linear |R| = 9.010147e-09
      9 Linear |R| = 8.988535e-09
     10 Linear |R| = 8.955818e-09
     11 Linear |R| = 8.893175e-09
     12 Linear |R| = 8.763566e-09
     13 Linear |R| = 8.504297e-09
     14 Linear |R| = 8.103937e-09
     15 Linear |R| = 7.454319e-09
     16 Linear |R| = 6.839006e-09
     17 Linear |R| = 5.880821e-09
     18 Linear |R| = 5.170450e-09
     19 Linear |R| = 4.222539e-09
     20 Linear |R| = 3.376157e-09
     21 Linear |R| = 2.587485e-09
     22 Linear |R| = 1.984959e-09
     23 Linear |R| = 1.508259e-09
     24 Linear |R| = 1.178323e-09
     25 Linear |R| = 9.336882e-10
     26 Linear |R| = 7.058346e-10
     27 Linear |R| = 5.551710e-10
     28 Linear |R| = 4.198611e-10
     29 Linear |R| = 3.374034e-10
     30 Linear |R| = 5.871209e-09
     31 Linear |R| = 5.847506e-09
     32 Linear |R| = 2.595568e-09
     33 Linear |R| = 2.313918e-09
     34 Linear |R| = 2.085521e-09
     35 Linear |R| = 1.767244e-09
     36 Linear |R| = 1.421870e-09
     37 Linear |R| = 1.051410e-09
     38 Linear |R| = 8.035187e-10
     39 Linear |R| = 5.871254e-10
     40 Linear |R| = 4.542527e-10
     41 Linear |R| = 3.245219e-10
     42 Linear |R| = 2.340989e-10
     43 Linear |R| = 1.600106e-10
     44 Linear |R| = 1.141013e-10
     45 Linear |R| = 8.264118e-11
     46 Linear |R| = 6.363237e-11
     47 Linear |R| = 4.606866e-11
     48 Linear |R| = 3.695673e-11
     49 Linear |R| = 3.015359e-11
     50 Linear |R| = 2.427008e-11
  Linear solve did not converge due to DIVERGED_ITS iterations 50
 2 Nonlinear |R| = 4.032112e-10
    |residual|_2 of individual variables:
                       w:    3.84205e-12
                       c:    2.47542e-11
                       eta1: 1.08331e-13
                       eta2: 6.64417e-17
                       eta3: 1.3091e-12
                       c1:   2.84151e-10
                       c2:   2.84007e-10
                       c3:   2.34102e-11
      0 Linear |R| = 4.032112e-10
      1 Linear |R| = 9.910058e-11
      2 Linear |R| = 2.129399e-11
      3 Linear |R| = 1.666190e-11
      4 Linear |R| = 1.177270e-11
      5 Linear |R| = 8.907619e-12
      6 Linear |R| = 6.862658e-12
      7 Linear |R| = 4.720409e-12
      8 Linear |R| = 3.422265e-12
      9 Linear |R| = 2.714312e-12
     10 Linear |R| = 2.103812e-12
     11 Linear |R| = 1.671415e-12
     12 Linear |R| = 1.394507e-12
     13 Linear |R| = 1.047202e-12
     14 Linear |R| = 8.452052e-13
     15 Linear |R| = 6.944958e-13
     16 Linear |R| = 5.150750e-13
     17 Linear |R| = 4.053526e-13
     18 Linear |R| = 3.429931e-13
     19 Linear |R| = 2.808270e-13
     20 Linear |R| = 2.169458e-13
     21 Linear |R| = 1.632714e-13
     22 Linear |R| = 1.290776e-13
     23 Linear |R| = 9.948311e-14
     24 Linear |R| = 7.628235e-14
     25 Linear |R| = 5.842547e-14
     26 Linear |R| = 4.751091e-14
     27 Linear |R| = 3.907655e-14
  Linear solve converged due to CONVERGED_RTOL iterations 27
 3 Nonlinear |R| = 7.619928e-13
    |residual|_2 of individual variables:
                       w:    4.22469e-15
                       c:    4.47435e-14
                       eta1: 1.57013e-16
                       eta2: 8.84063e-20
                       eta3: 2.382e-15
                       c1:   5.3707e-13
                       c2:   5.37337e-13
                       c3:   3.78424e-14
Nonlinear solve converged due to CONVERGED_FNORM_ABS iterations 3
 Solve Converged!

**Reply by aeslaughter** [reply]
Could the problem have reached a steady state?

**Reply by fdkong** [reply]
What are the initial residuals  for the first several steps?

 You could  use "`steady_state_detection=true`" to `Executioner` block, if the steady state is reached, no nonlinear solver will be involved

**Reply by LiuPengPeter** [reply]
The initial residuals for the first several steps are:

Steady-State Relative Differential Norm: 
0.000996499
1.3265e-05
1.33776e-06
1.02387e-06
6.08638e-07
5.03052e-07

**Comment by LiuPengPeter** [comment]
Do I need normalization to improve the converge?

**Reply by GiudGiud** [reply]
You could use scaling. Either manually scale the variables with a scaling = xx in the variable blocks or turn on automatic_scaling = true in the Executioner block
